iceoryx2-cal/src/event/common.rs

```diff
@@ -0,0 +1,523 @@
+                .open()
+            {
+                Ok(storage) => Ok(Notifier {
+                    storage,
+                    _tracker: PhantomData,
+                    _wait_mechanism: PhantomData,
+                }),
+                Err(DynamicStorageOpenError::DoesNotExist) => {
+                    fail!(from self, with NotifierCreateError::DoesNotExist,
+                        "{} since it does not exist.", msg);
+                }
+                Err(DynamicStorageOpenError::VersionMismatch) => {
+                    fail!(from self, with NotifierCreateError::VersionMismatch,
+                        "{} since the version of the existing construct does not match.", msg);
+                }
+                Err(DynamicStorageOpenError::InitializationNotYetFinalized) => {
+                    fail!(from self, with NotifierCreateError::InitializationNotYetFinalized,
+                        "{} since the initialization is after a timeout of {:?} still not finalized..",
+                        msg, self.creation_timeout);
+                }
+                Err(e) => {
+                    fail!(from self, with NotifierCreateError::InternalFailure,
+                        "{} due to an internal failure ({:?}).", msg, e);
+                }
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct Listener<
+        Tracker: IdTracker,
+        WaitMechanism: SignalMechanism,
+        Storage: DynamicStorage<Management<Tracker, WaitMechanism>>,
+    > {
+        storage: Storage,
+        _tracker: PhantomData<Tracker>,
+        _wait_mechanism: PhantomData<WaitMechanism>,
+    }
+
+    impl<
+            Tracker: IdTracker,
+            WaitMechanism: SignalMechanism,
+            Storage: DynamicStorage<Management<Tracker, WaitMechanism>>,
+        > NamedConcept for Listener<Tracker, WaitMechanism, Storage>
+    {
+        fn name(&self) -> &FileName {
+            self.storage.name()
+        }
+    }
+
+    impl<
+            Tracker: IdTracker,
+            WaitMechanism: SignalMechanism,
+            Storage: DynamicStorage<Management<Tracker, WaitMechanism>>,
+        > crate::event::Listener for Listener<Tracker, WaitMechanism, Storage>
+    {
+        fn try_wait(
+            &self,
+        ) -> Result<Option<crate::event::TriggerId>, crate::event::ListenerWaitError> {
+            while unsafe { self.storage.get().signal_mechanism.try_wait()? } {}
+            Ok(unsafe { self.storage.get().id_tracker.acquire() })
+        }
+
+        fn timed_wait(
+            &self,
+            timeout: Duration,
+        ) -> Result<Option<crate::event::TriggerId>, crate::event::ListenerWaitError> {
+            if let Some(id) = self.try_wait()? {
+                return Ok(Some(id));
+            }
+
+            if unsafe { self.storage.get().signal_mechanism.timed_wait(timeout)? } {
+                return Ok(unsafe { self.storage.get().id_tracker.acquire() });
+            }
+
+            Ok(None)
```

Comment:
How do you like something like this?
```suggestion
Ok(unsafe {
    self.storage
        .get()
        .signal_mechanism
        .timed_wait(timeout)?
        .then_some(self.storage.get().id_tracker.acquire())
        .flatten()
})
```